File: adapters/backend/ruby/lib/rampart/claims.rb
# frozen_string_literal: true

module Rampart
  # Verified JWT claims extracted from a Rampart access token.
  #
  # After successful authentication by {Rampart::Middleware}, an instance of
  # this class is stored in +env['rampart.claims']+.
  class Claims
    ATTRIBUTES = %i[
      iss sub iat exp org_id preferred_username
      email email_verified given_name family_name roles
    ].freeze

    attr_reader(*ATTRIBUTES)

    # @param payload [Hash] Decoded JWT payload (string keys).
    def initialize(payload)
      @iss                = payload["iss"]
      @sub                = payload["sub"]
      @iat                = payload["iat"]
      @exp                = payload["exp"]
      @org_id             = payload["org_id"]
      @preferred_username = payload["preferred_username"]
      @email              = payload["email"]
      @email_verified     = !!payload["email_verified"]
      @given_name         = payload["given_name"]
      @family_name        = payload["family_name"]
      @roles              = Array(payload["roles"]).map(&:to_s)
    end

    # Returns the claims as a plain Hash with string keys.
    def to_h
      ATTRIBUTES.each_with_object({}) do |attr, hash|
        hash[attr.to_s] = public_send(attr)
      end
    end
  end
end

File: adapters/backend/rust/src/claims.rs
//! JWT claims extracted from Rampart access tokens.

use serde::{Deserialize, Serialize};

/// Verified JWT claims from a Rampart access token.
///
/// Fields map to the JSON claim names used by the Rampart IAM server.
/// Optional fields (`given_name`, `family_name`, `roles`) default to
/// empty values when absent from the token.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Claims {
    /// Subject identifier (user UUID) — JWT "sub" claim.
    pub sub: String,

    /// Issuer URL — JWT "iss" claim.
    pub iss: String,

    /// Token issued-at time as a Unix timestamp — JWT "iat" claim.
    pub iat: f64,

    /// Token expiration time as a Unix timestamp — JWT "exp" claim.
    pub exp: f64,

    /// Organization UUID the user belongs to — custom "org_id" claim.
    #[serde(default)]
    pub org_id: String,

    /// User's display name — OIDC "preferred_username" claim.
    #[serde(default)]
    pub preferred_username: String,

    /// User's email address — OIDC "email" claim.
    #[serde(default)]
    pub email: String,

    /// Whether the email address has been verified — OIDC "email_verified" claim.
    #[serde(default)]
    pub email_verified: bool,

    /// User's first name — OIDC "given_name" claim.
    #[serde(default)]
    pub given_name: String,

    /// User's last name — OIDC "family_name" claim.
    #[serde(default)]
    pub family_name: String,

    /// Roles assigned to the user — custom "roles" claim.
    #[serde(default)]
    pub roles: Vec<String>,
}

impl Claims {
    /// Returns `true` if the user has the specified role.
    pub fn has_role(&self, role: &str) -> bool {
        self.roles.iter().any(|r| r == role)
    }

    /// Returns `true` if the user has all of the specified roles.
    pub fn has_all_roles(&self, roles: &[&str]) -> bool {
        roles.iter().all(|r| self.has_role(r))
    }

    /// Returns the list of roles from `required` that the user is missing.
    pub fn missing_roles<'a>(&self, required: &[&'a str]) -> Vec<&'a str> {
        required
            .iter()
            .filter(|r| !self.has_role(r))
            .copied()
            .collect()
    }
}
